In [52]:

import numpy as np
from scipy.stats import geom

p = 0.3     # Probabilidad de éxito
N = 1000    # Tamaño de muestra

muestra_teorica = geom.rvs(p, size=N)


In [53]:
U = np.random.rand(N)
muestra_empirica = np.ceil(np.log(1-U) / np.log(1-p))


In [54]:
import numpy as np
from scipy.stats import chisquare

# Definir clases (valores posibles)
valores = np.arange(1, max(muestra_teorica.max(), muestra_empirica.max()) + 1)

# Frecuencias observadas y esperadas
freq_teo, _ = np.histogram(muestra_teorica, bins=np.arange(1, valores[-1]+2))
freq_emp, _ = np.histogram(muestra_empirica, bins=np.arange(1, valores[-1]+2))

# Filtrar frecuencias donde f_exp (frecuencia esperada) sea 0 para evitar problemas de división
valid_bins = freq_teo > 0  # Solo mantener las categorías con frecuencias esperadas > 0

freq_teo_filtrado = freq_teo[valid_bins]
freq_emp_filtrado = freq_emp[valid_bins]

# Normalizar las frecuencias para que sumen lo mismo
sum_freq_emp = np.sum(freq_emp_filtrado)
sum_freq_teo = np.sum(freq_teo_filtrado)

# Ajustar las frecuencias para que la suma sea igual
freq_emp_normalized = freq_emp_filtrado * (sum_freq_teo / sum_freq_emp)
freq_teo_normalized = freq_teo_filtrado  # Ya debería sumar lo mismo si las normalizamos

# Aplicar la prueba Chi-Cuadrado con las frecuencias normalizadas
chi2, p_chi = chisquare(freq_emp_normalized, f_exp=freq_teo_normalized)

print(f"Chi2 = {chi2:.4f}, p-valor = {p_chi:.4f}")


Chi2 = 31.0499, p-valor = 0.0284


In [55]:
from scipy.stats import ks_2samp

ks_stat, p_ks = ks_2samp(muestra_teorica, muestra_empirica)

print(f"KS = {ks_stat:.4f}, p-valor = {p_ks:.4f}")


KS = 0.0250, p-valor = 0.9137


Las pruebas realizadas sobre las muestras generadas de una distribución geométrica  
muestran resultados contradictorios. La prueba de Chi-Cuadrado indica que las muestras no provienen de la misma distribución, ya que el p-valor es 0.0284, lo que sugiere una diferencia significativa. Sin embargo, la prueba de Kolmogorov-Smirnov no rechaza la hipótesis nula, ya que el p-valor es 0.9137, indicando que no hay diferencias significativas entre las distribuciones acumuladas. 